chain/sub/incoming.go

```diff
@@ -121,34 +144,192 @@ func (bv *BlockValidator) flagPeer(p peer.ID) {
 
 func (bv *BlockValidator) Validate(ctx context.Context, pid peer.ID, msg *pubsub.Message) pubsub.ValidationResult {
 	stats.Record(ctx, metrics.BlockReceived.M(1))
+
+	recordFailure := func(what string) {
+		ctx, _ = tag.New(ctx, tag.Insert(metrics.FailureType, what))
+		stats.Record(ctx, metrics.BlockValidationFailure.M(1))
+		bv.flagPeer(pid)
+	}
+
+	// make sure the block can be decoded
 	blk, err := types.DecodeBlockMsg(msg.GetData())
 	if err != nil {
 		log.Error("got invalid block over pubsub: ", err)
-		ctx, _ = tag.New(ctx, tag.Insert(metrics.FailureType, "invalid"))
-		stats.Record(ctx, metrics.BlockValidationFailure.M(1))
-		bv.flagPeer(pid)
+		recordFailure("invalid")
 		return pubsub.ValidationReject
 	}
 
+	// check the message limit constraints
 	if len(blk.BlsMessages)+len(blk.SecpkMessages) > build.BlockMessageLimit {
 		log.Warnf("received block with too many messages over pubsub")
-		ctx, _ = tag.New(ctx, tag.Insert(metrics.FailureType, "too_many_messages"))
-		stats.Record(ctx, metrics.BlockValidationFailure.M(1))
-		bv.flagPeer(pid)
+		recordFailure("too_many_messages")
 		return pubsub.ValidationReject
 	}
 
+	// make sure we have a signature
+	if blk.Header.BlockSig == nil {
+		log.Warnf("received block without a signature over pubsub")
+		recordFailure("missing_signature")
+		return pubsub.ValidationReject
+	}
+
+	// validate the block meta: the Message CID in the header must match the included messages
+	err = bv.validateMsgMeta(ctx, blk)
+	if err != nil {
+		log.Warnf("error validating message metadata: %s", err)
+		recordFailure("invalid_block_meta")
+		return pubsub.ValidationReject
+	}
+
+	// we want to ensure that it is a block from a known miner; we reject blocks from unknown miners
+	// to prevent spam attacks.
+	// the logic works as follows: we lookup the miner in the chain for its key.
+	// if we can find it then it's a known miner and we can validate the signature.
+	// if we can't find it, we check whether we are (near) synced in the chain.
+	// if we are not synced we cannot validate the block and we must ignore it.
+	// if we are synced and the miner is unknown, then the block is rejcected.
+	// TODO verify that the actor power is > 0
+	key, err := bv.getMinerWorkerKey(ctx, blk)
+	if err != nil {
+		if bv.isChainNearSynced() {
+			log.Warnf("received block message from unknown miner over pubsub; rejecting message")
+			recordFailure("unknown_miner")
+			return pubsub.ValidationReject
+		} else {
+			log.Warnf("cannot validate block message; unknown miner in unsynced chain")
+			return pubsub.ValidationIgnore
+		}
+	}
+
+	blob, err := blk.Header.SigningBytes()
+	if err != nil {
+		// this shouldn't happen; it's an encoding error of the object we just decoded sans the
+		// signature...
+		log.Errorf("error retrieving signing bytes for block header: %s", err)
+		return pubsub.ValidationIgnore
+	}
+
+	err = sigs.Verify(blk.Header.BlockSig, key, blob)
```

Comment:
since bls signatures are rather expensive, we want to make sure we're not validating the same signature twice. in the chain sync code we do this in blockSigCheck (actually looking at that, theres a nice helper function for us `sigs.CheckBlockSignature`)

Either we can *remove* that call entirely (which feels risky? maybe?) or we can cache which signatures we have validated somewhere

Reply:
We shouldn't remove that, it feels very bad to accept blocks without validating signatures.

Regardless, I will use the friendly utility function for the validation.
How should we go about caching validated signatures?

Reply:
After sync discussion, we don't think that the 5-6ms of signature validation time is going to materially affect the validator.
Nonetheless, It would still be nice to avoid validating it twice further downstream -- so I will add a TODO.

Reply:
Actually it is not that hard to do, provided we consistently use `CheckBlockSignature`.
We can add a non-exported field to the block header, with associated methods, and have `CheckBlockSignature` cache the fact that it has been validated and short-circuit the second time it's called.

Reply:
done.